### PR TITLE
xcatjktest change (update xcatjktest to run xcat installation as an individual case)

### PR DESCRIPTION
**Pull request by tingtli**
@hu-weihua  since #1670 is messy. I could like to open a new PR to trace all the changes

### CODE REVIEW

File: xCAT-server/share/xcat/tools/jenkins/xcatjktest [hunk 1]
```diff
@@ -343,7 +343,13 @@ sub get_build{
    
     &runcmd("cd $xcatpackagesdir >/dev/null 2>&1 && wget $xcatdep_addr >/dev/null 2>&1");
     if($?){
-        send_msg(0, "[get_build] Can't download xcat-dep from $xcatcore_addr");
+        send_msg(0, "[get_build] Can't download xcat-dep from $xcatdep_addr");
+        return 1;
+    }
+
+    &runcmd("cd $xcatpackagesdir >/dev/null 2>&1 && wget http://xcat.org/files/go-xcat >/dev/null 2>&1");
```
Comment (hu-weihua) [inline xCAT-server/share/xcat/tools/jenkins/xcatjktest]: Now we support clone ``xcatjktest`` for each project and each time running, so after cloning, the ``/TEST/$JOB_NAME-$BUILD_NUMBER/xcatjktest`` directory is unique for one project one time running. So the initialization of attribute ``$xcatpackagesdir`` isn't suitable now

     254 sub init{
      ....
     259     $xcatpackagesdir="$homedir/xcatpackages/$proname/$proruntime";

The last 2 levels directory is useless now.

File: xCAT-server/share/xcat/tools/jenkins/xcatjktest [hunk 2]
```diff
@@ -834,15 +845,28 @@ sub install_xcat {
 
     &runcmd("scp $mn:/$installlog $logfiledir/new_xcat_installation.log >/dev/null 2>&1"); 
 
-    #check if MN is installed successful
-    my $lsdefout = `ssh -t $mn 'bash -l -i -c "lsdef -v"'`;
-    chomp($lsdefout);
-    if($lsdefout !~ /^lsdef - Version/ ){
-        send_msg(0, "[install_xcat] install xcat in $mn failed");
-        return 1;
+    #check if xcat-test is installed successfully on mn
```
Comment (hu-weihua) [inline xCAT-server/share/xcat/tools/jenkins/xcatjktest]: Could you update line 846? now the log file should be ``new_xcattest_installation.log ``

File: xCAT-server/share/xcat/tools/jenkins/xcatjktest [hunk 3]
```diff
@@ -898,6 +923,37 @@ sub do_test {
         }
         close(FILE);
         &runcmd("echo $casecnt > /tmp/totalcase.$proid");
+
+      #run the special install_xcat case first
+      my $rst1=0;
+      my @output1;
+      send_msg(2, "[do_test] doing test for case install_xcat in $mn.....");
+      if($os !~ /ubuntu/i){
+         @output1=runcmd("xdsh $mn \"xcattest -f /opt/xcat/share/xcat/tools/autotest/default.conf:System  -t install_xcat\" >/dev/null 2>&1");
```
Comment (hu-weihua) [inline xCAT-server/share/xcat/tools/jenkins/xcatjktest]: If we throw the output of ``xdsh`` to ``/dev/null``, ``@output1`` will be empty always. I think these code won't work as what we expected . Maybe we can refine code like below
 
    @output1=runcmd("xdsh $mn \"xcattest -f /opt/xcat/share/xcat/tools/autotest/default.conf:System  -t install_xcat  2>&1 \"");

Due to sub function ``runcmd`` has handled the ``2>&1``, we don't have to do it here.

Could you please update line935, line 962 and line 965 at the same time? they have the same problem.

File: xCAT-server/share/xcat/tools/jenkins/xcatjktest [hunk 4]
```diff
@@ -898,6 +923,37 @@ sub do_test {
...
+         $rst1=$?;
+       }else{
+          @output1=runcmd("ssh -t $mn 'exec bash -l -i -c \"xcattest -f /opt/xcat/share/xcat/tools/autotest/default.conf:System -t install_xcat\"' >/dev/null 2>&1");
+          $rst1=$?;
+       }
+       my $tmpoutput1=join(' ', @output1);
+       send_msg(2, "[do_test] install_xcat output of trigger xcattest: $tmpoutput1");
```
Comment (hu-weihua) [inline xCAT-server/share/xcat/tools/jenkins/xcatjktest]: I think we just need to log the output into log file when line 932/935 failed.

File: xCAT-server/share/xcat/tools/jenkins/xcatjktest [hunk 5]
```diff
@@ -898,15 +923,46 @@ sub do_test {
...
+      send_msg(2, "[do_test] doing test for case install_xcat in $mn.....");
+      if($os !~ /ubuntu/i){
+         @output1=runcmd("xdsh $mn \"xcattest -f /opt/xcat/share/xcat/tools/autotest/default.conf:System  -t install_xcat  2>&1 \"");
+         $rst1=$?;
+       }else{
+          @output1=runcmd("ssh -t $mn 'exec bash -l -i -c \"xcattest -f /opt/xcat/share/xcat/tools/autotest/default.conf:System -t install_xcat 2>&1\"'");
+          $rst1=$?;
+       }
+       my $tmpoutput1=join(' ', @output1);
+       send_msg(2, "[do_test] install_xcat output of trigger xcattest: $tmpoutput1");
```
Comment (hu-weihua) [inline xCAT-server/share/xcat/tools/jenkins/xcatjktest]: I think we just need to log the output into log file when line 932/935 failed

File: xCAT-server/share/xcat/tools/jenkins/xcatjktest [hunk 6]
```diff
@@ -898,15 +923,46 @@ sub do_test {
         }
         close(FILE);
         &runcmd("echo $casecnt > /tmp/totalcase.$proid");
+
+      #run the special install_xcat case first
+      my $rst1=0;
+      my @output1;
+      send_msg(2, "[do_test] doing test for case install_xcat in $mn.....");
+      if($os !~ /ubuntu/i){
+         @output1=runcmd("xdsh $mn \"xcattest -f /opt/xcat/share/xcat/tools/autotest/default.conf:System  -t install_xcat  2>&1 \"");
```
Comment (hu-weihua) [inline xCAT-server/share/xcat/tools/jenkins/xcatjktest]: I saw line 938-939, that means no matter line 932 was successful or not, the output of line 932 will be logged into log file anyway. So if we don't  throw the output of ``xdsh`` to ``/dev/null``, there are so many log output by ``xcattest`` will be logged, it's terrible.

So I think we should use ``/opt/xcat/share/xcat/tools/autotest/result/*`` to cover the error of ``xcattest``, using ``@outptut`` to cover the error of ``xdsh``. 

Maybe we could refine the code like below

     @output1=runcmd("xdsh $mn \"xcattest -f /opt/xcat/share/xcat/tools/autotest/default.conf:System  -t install_xcat   >/dev/null \"");

Due to sub function ``runcmd`` has handled the ``2>&1``, we don't have to do it here.

Could you please update line935, line 962 and line 965 at the same time? they have the same problem.

File: xCAT-server/share/xcat/tools/jenkins/xcatjktest [hunk 7]
```diff
@@ -898,15 +923,47 @@ sub do_test {
         }
         close(FILE);
         &runcmd("echo $casecnt > /tmp/totalcase.$proid");
+
+      #run the special install_xcat case first
+      my $rst1=0;
+      my @output1;
+      send_msg(2, "[do_test] doing test for case install_xcat in $mn.....");
+      if($os !~ /ubuntu/i){
+         @output1=runcmd("xdsh $mn \"xcattest -f /opt/xcat/share/xcat/tools/autotest/default.conf:System  -t install_xcat >/dev/null\"");
```
Comment (caomengmeng) [inline xCAT-server/share/xcat/tools/jenkins/xcatjktest]: @tingtli  it is install_xCAT_on_rhels_sles now after reviewed.

File: xCAT-server/share/xcat/tools/jenkins/xcatjktest [hunk 8]
```diff
@@ -898,15 +923,47 @@ sub do_test {
...
+         $rst1=$?;
+       }else{
+         @output1=runcmd("ssh -t $mn 'exec bash -l -i -c \"xcattest -f /opt/xcat/share/xcat/tools/autotest/default.conf:System -t install_xcat >/dev/null\"'");
```
Comment (caomengmeng) [inline xCAT-server/share/xcat/tools/jenkins/xcatjktest]: @tingtli  it is install_xCAT_on_ubuntu now after reviewed.

File: xCAT-server/share/xcat/tools/jenkins/xcatjktest [hunk 9]
```diff
@@ -815,7 +815,7 @@ sub install_xcattest {
             &runcmd("xdsh $mn \"/xcat-dep/mklocalrepo.sh\" >/dev/null 2>&1");
             &runcmd("xdsh $mn 'apt-get clean all' >/dev/null 2>&1");
             &runcmd("xdsh $mn 'apt-get update' >/dev/null 2>&1");
-            &runcmd("xdsh $mn 'apt-get -y install xcat-test  > $installlog 2>&1' >/dev/null 2>&1");
+            &runcmd("xdsh $mn \"cd /xcat-core/pool/main/x/xcat-test,dpkg -i xcat-test_*.deb > $installlog 2>&1\" >/dev/null 2>&1"); 
         }elsif($arch =~ /le/i || $arch =~ /el/i){
```
Comment (hu-weihua) [inline xCAT-server/share/xcat/tools/jenkins/xcatjktest]: I think below command is simpler :

     &runcmd("xdsh $mn \"dpkg -i /xcat-core/pool/main/x/xcat-test/xcat-test_*.deb > $installlog 2>&1\" >/dev/null 2>&1");

File: xCAT-server/share/xcat/tools/jenkins/xcatjktest [hunk 10]
```diff
@@ -724,7 +724,7 @@ sub install_xcattest {
             &runcmd("xdsh $mn \"cd /xcat-dep/rh$version/$arch && ./mklocalrepo.sh\" >/dev/null 2>&1");
         }
         &runcmd("xdsh $mn \"rpm --import /RPM-GPG-KEY-redhat-release\" >/dev/null 2>&1");
-        &runcmd("xdsh $mn \"yum -y install xCAT-test > $installlog 2>&1\" >/dev/null 2>&1");
+        &runcmd("xdsh $mn \"cd /xcat-core && rpm -ivh xCAT-test-*.rpm --nodeps\" >/dev/null 2>&1");
         &runcmd("xdsh $mn \"yum -y install createrepo expect\" >/dev/null 2>&1");
```
Comment (hu-weihua) [inline xCAT-server/share/xcat/tools/jenkins/xcatjktest]: I think below command is simpler, and add `` > $installlog 2>&1``

    &runcmd("xdsh $mn \"rpm -ivh /xcat-core/xCAT-test-*.rpm --nodeps > $installlog 2>&1 \" >/dev/null 2>&1");

File: xCAT-server/share/xcat/tools/jenkins/xcatjktest [hunk 11]
```diff
@@ -747,7 +747,7 @@ sub install_xcattest {
         &runcmd("xdsh $mn \"zypper ar file:///xcat-core xCAT-core\" >/dev/null 2>&1");
         &runcmd("xdsh $mn \"zypper sl -U\" >/dev/null 2>&1");
         &runcmd("xdsh $mn \"zypper --gpg-auto-import-keys search --match-exact -s screen\" >/dev/null 2>&1");
-        &runcmd("xdsh $mn \"zypper -n install xCAT-test > $installlog 2>&1\" >/dev/null 2>&1");
+        &runcmd("xdsh $mn \"cd /xcat-core && rpm -ivh xCAT-test-*.rpm --nodeps\" >/dev/null 2>&1");
         &runcmd("xdsh $mn \"zypper -n install createrepo expect\" >/dev/null 2>&1");
```
Comment (hu-weihua) [inline xCAT-server/share/xcat/tools/jenkins/xcatjktest]: Same comment with line727

**Pull request merged**